stringdtype/stringdtype/src/casts.c

```diff
@@ -64,12 +65,293 @@ PyArrayMethod_Spec StringToStringCastSpec = {
+{
+    if (code <= 0x7F) {
+        // 0zzzzzzz -> 0zzzzzzz
+        c[0] = (char)code;
+        return 1;
+    }
+    else if (code <= 0x07FF) {
+        // 00000yyy yyzzzzzz -> 110yyyyy 10zzzzzz
+        c[0] = (0xC0 | (code >> 6));
+        c[1] = (0x80 | (code & 0x3F));
+        return 2;
+    }
+    else if (code <= 0xFFFF) {
+        // xxxxyyyy yyzzzzzz -> 110yyyyy 10zzzzzz
+        c[0] = (0xe0 | (code >> 12));
+        c[1] = (0x80 | ((code >> 6) & 0x3f));
+        c[2] = (0x80 | (code & 0x3f));
+        return 3;
+    }
+    else if (code <= 0x10FFFF) {
+        // 00wwwxx xxxxyyyy yyzzzzzz -> 11110www 10xxxxxx 10yyyyyy 10zzzzzz
+        c[0] = (0xf0 | (code >> 18));
+        c[1] = (0x80 | ((code >> 12) & 0x3f));
+        c[2] = (0x80 | ((code >> 6) & 0x3f));
+        c[3] = (0x80 | (code & 0x3f));
+        return 4;
+    }
+    return -1;
+}
+
+static int
+unicode_to_string(PyArrayMethod_Context *context, char *const data[],
+                  npy_intp const dimensions[], npy_intp const strides[],
+                  NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr **descrs = context->descriptors;
+    long in_size = (descrs[0]->elsize) / 4;
+
+    npy_intp N = dimensions[0];
+    Py_UCS4 *in = (Py_UCS4 *)data[0];
+    char **out = (char **)data[1];
+
+    // 4 bytes per UCS4 character
+    npy_intp in_stride = strides[0] / 4;
+    // strides are in bytes but pointer offsets are in pointer widths, so
+    // divide by the element size (one pointer width) to get the pointer offset
+    npy_intp out_stride = strides[1] / context->descriptors[1]->elsize;
+
+    while (N--) {
+        // pessimistically allocate 4 bytes per allowed character
+        // plus one byte for the null terminator
+        char *out_buf = malloc((in_size * 4 + 1) * sizeof(char));
+        size_t out_num_bytes = 0;
+        for (int i = 0; i < in_size; i++) {
+            // get code point
+            Py_UCS4 code = in[i];
+
+            if (code == 0) {
+                break;
+            }
+
+            // convert codepoint to UTF8 bytes
+            char utf8_c[4] = {0};
+            size_t num_bytes = ucs4_code_to_utf8_char(code, utf8_c);
+            out_num_bytes += num_bytes;
+
+            if (num_bytes == -1) {
+                // acquire GIL, set error, return
+                PyGILState_STATE gstate;
+                gstate = PyGILState_Ensure();
+                PyErr_SetString(PyExc_TypeError,
+                                "Invalid unicode code point found");
+                PyGILState_Release(gstate);
+                return -1;
+            }
+
+            // copy utf8_c into out_buf
+            strncpy(out_buf, utf8_c, num_bytes);
+
+            // increment out_buf by the size of the character
+            out_buf += num_bytes;
+        }
+
+        // reset out_buf to the beginning of the string
+        out_buf -= out_num_bytes;
+
+        // pad string with null character
+        out_buf[out_num_bytes] = '\0';
```

Comment:
For NumPy you have to pad, for your strings not ;).  I guess you don't currently have the length of the string stored at the beginning of each?  Sounds like a good addition, but also maybe something for later...

Reply:
Yeah I was thinking today about whether we should store a struct or use some other trick to keep the length and completely avoid the possibility of buffer overruns.

Reply:
I guess something like `{int length; char *buf[]}` would be the base struct.  Then you just allocate the size with `sizeof(struct) + length_of_string` and can cast/access this nicely.
I guess you could do tricks to safe storage on short strings, but dunno if that is worthwhile (basically, the first unicode character could actually be the length).

Reply:
It’s a lot of code to vendor and we probably don’t need most of the features, but `sds`[https://github.com/antirez/sds#sds-internals-and-advanced-usage] might be worth looking at, in particular the way the library simply returns a `char *` but uses [a trick](https://github.com/antirez/sds#sds-internals-and-advanced-usage) to ensure there’s always a valid header before the string.

Reply:
Ah, they have a free also to allow growing.  I guess storing the actual start can nice in practice.  If it is useful, vendoring seems fine to me (or submodule)?

I suppose we can invent an arbitrary amount of nifty storage schemes :), if you store pointer+length, then you have at least 10 bytes for short strings to store without any allocation ;).

Reply:
I also like the idea of having a struct of `{int length; char *buf[]}`, I think it would make the logic here nicer. If vendoring seems best to everyone else here, I'm all for it :)